contentcuration/contentcuration/models.py

```diff
@@ -170,13 +214,166 @@ class Meta:
+
+    def save(self, *args, **kwargs):
+        """
+        Overrider the default save method.
+        If the content_copy FileField gets passed a content copy:
+            1. generate the MD5 from the content copy
+            2. fill the other fields accordingly
+            3. update tracking for this content copy
+        If None is passed to the content_copy FileField:
+            1. delete the content copy.
+            2. update tracking for this content copy
+        """
+        if self.content_copy:  # if content_copy is supplied, hash out the file
+            md5 = hashlib.md5()
+            for chunk in self.content_copy.chunks():
+                md5.update(chunk)
+
+            self.checksum = md5.hexdigest()
+            self.file_size = self.content_copy.size
+            self.extension = os.path.splitext(self.content_copy.name)[1]
+            # update ContentCopyTracking
+            try:
+                content_copy_track = ContentCopyTracking.objects.get(content_copy_id=self.checksum)
+                content_copy_track.referenced_count += 1
+                content_copy_track.save()
+            except ContentCopyTracking.DoesNotExist:
+                ContentCopyTracking.objects.create(referenced_count=1, content_copy_id=self.checksum)
+        else:
+            # update ContentCopyTracking, if referenced_count reach 0, delete the content copy on disk
+            try:
+                content_copy_track = ContentCopyTracking.objects.get(content_copy_id=self.checksum)
+                content_copy_track.referenced_count -= 1
+                content_copy_track.save()
+                if content_copy_track.referenced_count == 0:
+                    content_copy_path = os.path.join(settings.CONTENT_COPY_DIR, self.checksum[0:1], self.checksum[1:2], self.checksum + self.extension)
+                    if os.path.isfile(content_copy_path):
+                        os.remove(content_copy_path)
+            except ContentCopyTracking.DoesNotExist:
+                pass
+            self.checksum = None
+            self.file_size = None
+            self.extension = None
+        super(File, self).save(*args, **kwargs)
+
+class License(models.Model):
+    """
+    Normalize the license of ContentNode model
+    """
+    license_name = models.CharField(max_length=50)
     exists = models.BooleanField(
         default=False,
         verbose_name=_("license exists"),
         help_text=_("Tells whether or not a content item is licensed to share"),
     )
 
+    def __str__(self):
+        return self.license_name
+
+class PrerequisiteContentRelationship(models.Model):
+    """
+    Predefine the prerequisite relationship between two ContentNode objects.
+    """
+    contentnode_1 = models.ForeignKey(ContentNode, related_name='%(app_label)s_%(class)s_1')
+    contentnode_2 = models.ForeignKey(ContentNode, related_name='%(app_label)s_%(class)s_2')
```

Comment:
I changed the field names to make more sense in Kolibri repo, I should change them here too to keep the consistency.